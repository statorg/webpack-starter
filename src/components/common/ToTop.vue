<template>
  <transition name="to-top">
    <button
      v-show="visible"
      class="to-top"
      type="button"
      aria-label="Scroll to top"
      data-fixed="margin"
      @click="onClick"
    >
      <Icon name="house" class="to-top__icon" />
    </button>
  </transition>
</template>

<script>
import Icon from './Icon'

export default {
  name: 'ToTop',
  components: { Icon },
  data () {
    return {
      offset: 600,
      scrollTop: 0
    }
  },
  computed: {
    visible () {
      return this.scrollTop > this.offset
    }
  },
  mounted () {
    window.addEventListener('scroll', this.onScroll)
  },
  beforeDestroy () {
    window.removeEventListener('scroll', this.onScroll)
  },
  methods: {
    /* istanbul ignore next */
    onScroll () {
      this.scrollTop = window.pageYOffset
    },
    onClick () {
      this.scrollToTop()
    },
    /* istanbul ignore next */
    scrollToTop () {
      const currentScroll = window.pageYOffset
      if (currentScroll > 0) {
        window.requestAnimationFrame(this.scrollToTop)
        window.scrollTo(0, Math.floor(currentScroll - (currentScroll / 5)))
      }
    }
  }
}
</script>

<style lang="scss">
.to-top {
  font-size: 16px;
  width: 60px;
  height: 60px;
  background-color: $primary;
  border-radius: 50%;
  position: fixed;
  border: none;
  right: 40px;
  bottom: 40px;
  z-index: z(sticky) - 1;
  transition: opacity $transition;
  cursor: pointer;
  color: #ffffff;

  @include down(md) {
    width: 55px;
    height: 55px;
    right: 20px;
    bottom: 20px;
  }

  &.to-top-enter,
  &.to-top-leave-to {
    opacity: 0;
  }
}
</style>
