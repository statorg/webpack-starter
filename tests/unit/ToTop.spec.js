import { shallowMount } from '@vue/test-utils'
import ToTop from '@/components/common/ToTop'

describe('ToTop.vue', () => {
  it('should show when scrollTop greater then offset', () => {
    const wrapper = shallowMount(ToTop)
    wrapper.setData({ scrollTop: wrapper.vm.offset + 1 })
    expect(wrapper.isVisible()).toBe(true)
  })

  it('should hide when button is clicked', () => {
    const wrapper = shallowMount(ToTop, {
      attachToDocument: true
    })
    wrapper.setMethods({
      scrollToTop () {
        this.scrollTop = 0
      }
    })
    wrapper.setData({ scrollTop: wrapper.vm.offset + 1 })
    expect(wrapper.isVisible()).toBe(true)
    wrapper.trigger('click')
    expect(wrapper.isVisible()).toBe(false)
  })

  it('should render and match snapshot', () => {
    const wrapper = shallowMount(ToTop)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
