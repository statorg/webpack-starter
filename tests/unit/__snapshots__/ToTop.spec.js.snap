// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToTop.vue should render and match snapshot 1`] = `
<button type="button" aria-label="Scroll to top" data-fixed="margin" class="to-top" style="display: none;" name="to-top">
  <icon-stub name="house" class="to-top__icon"></icon-stub>
</button>
`;
